### PR TITLE
native: add Dgehd2

### PR DESCRIPTION
**Pull request by vladimir-ch**

### CODE REVIEW

File: native/dgehd2.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+//  [     v1  h   h   h   h   h ]
+//  [     v1  v2  h   h   h   h ]
+//  [     v1  v2  v3  h   h   h ]
+//  [                         a ]
+// where a denotes an element of the original matrix A, h denotes a
+// modified element of the upper Hessenberg matrix H, and vi denotes an
+// element of the vector defining H_i.
+//
+// ilo and ihi determine the block of A that will be reduced to upper Hessenberg
+// form. It must hold that 0 <= ilo <= ihi <= max(0, n-1), otherwise Dgehd2 will
```
Comment (vladimir-ch) [inline native/dgehd2.go]: The reference implementation specifies special values if n == 0 but I'm not sure how important these are (probably plays some role in the correct input checks). As always, opinions and suggestions welcome.

Comment (vladimir-ch) [inline native/dgehd2.go]: Oops, sorry, I confused it with Dgehrd.

File: native/dgehd2.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+		panic("lapack: invalid value of ilo")
+	case ihi < min(ilo, n-1) || ihi >= n:
+		panic("lapack: invalid value of ihi")
+	case len(tau) != n-1:
+		panic(badTau)
+	case len(work) < n:
+		panic(badWork)
+	}
+
+	for i := ilo; i <= ihi-1; i++ {
```
Comment (kortschak) [inline native/dgehd2.go]: `i < ihi`?

File: native/dgehd2.go (hunk 1 above)
Comment (kortschak) [inline native/dgehd2.go]: Thanks, I was a little confused what I was looking for.

File: native/dgehd2.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+		// Compute elementary reflector H_i to annihilate A[i+2:ihi+1,i].
+		var aii float64
+		aii, tau[i] = impl.Dlarfg(ihi-i, a[(i+1)*lda+i], a[min(i+2, n-1)*lda+i:], lda)
+		a[(i+1)*lda+i] = 1
+
+		// Apply H_i to A[0:ihi+1,i+1:ihi+1] from the right.
+		impl.Dlarf(blas.Right, ihi+1, ihi-i, a[(i+1)*lda+i:], lda, tau[i], a[i+1:], lda, work)
```
Comment (kortschak) [inline native/dgehd2.go]: Why is this `ihi+1`?

Comment (vladimir-ch) [inline native/dgehd2.go]: It's the number of rows but `ihi` is 0-indexed ...

Comment (kortschak) [inline native/dgehd2.go]: Thanks.

File: native/dgehd2.go (hunk 2 above)
Comment (vladimir-ch) [inline native/dgehd2.go]: Done.

**Comment by btracey** [comment]
LGTM

**Pull request merged**